transports/quic/src/lib.rs

```diff
@@ -0,0 +1,653 @@
+use libp2p_core::{
+    identity::Keypair,
+    multiaddr::{Multiaddr, Protocol},
+    muxing::StreamMuxerEvent,
+    transport::{ListenerId, TransportError, TransportEvent},
+    PeerId, StreamMuxer, Transport,
+};
+
+use libp2p_tls as tls;
+
+use std::{
+    future::Future,
+    io,
+    net::{Ipv4Addr, Ipv6Addr, SocketAddr},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{future::BoxFuture, FutureExt};
+use futures::{stream::SelectAll, AsyncRead, AsyncWrite, Stream, StreamExt};
+
+mod in_addr;
+
+use in_addr::InAddr;
+
+pub struct QuicSubstream {
+    send: quinn::SendStream,
+    recv: quinn::RecvStream,
+    closed: bool,
+}
+
+impl QuicSubstream {
+    fn new(send: quinn::SendStream, recv: quinn::RecvStream) -> Self {
+        Self {
+            send,
+            recv,
+            closed: false,
+        }
+    }
+}
+
+impl AsyncRead for QuicSubstream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        AsyncRead::poll_read(Pin::new(&mut self.get_mut().recv), cx, buf)
+    }
+}
+
+impl AsyncWrite for QuicSubstream {
+    fn poll_write(self: Pin<&mut Self>, cx: &mut Context, buf: &[u8]) -> Poll<io::Result<usize>> {
+        AsyncWrite::poll_write(Pin::new(&mut self.get_mut().send), cx, buf)
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        AsyncWrite::poll_flush(Pin::new(&mut self.get_mut().send), cx)
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        let this = self.get_mut();
+        if this.closed {
+            // For some reason poll_close needs to be 'fuse'able
+            return Poll::Ready(Ok(()));
+        }
+        let close_result = AsyncWrite::poll_close(Pin::new(&mut this.send), cx);
+        if close_result.is_ready() {
+            this.closed = true;
+        }
+        close_result
+    }
+}
+
+pub struct QuicMuxer {
+    connection: quinn::Connection,
+    incoming:
+        BoxFuture<'static, Result<(quinn::SendStream, quinn::RecvStream), quinn::ConnectionError>>,
+    outgoing:
+        BoxFuture<'static, Result<(quinn::SendStream, quinn::RecvStream), quinn::ConnectionError>>,
+}
+
+impl StreamMuxer for QuicMuxer {
+    type Substream = QuicSubstream;
+    type Error = quinn::ConnectionError;
+
+    fn poll_inbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let this = self.get_mut();
+
+        let substream = futures::ready!(this.incoming.poll_unpin(cx));
```

Comment:
You could `?` right here. We close the connection as soon as any of these functions fail so there is no point in setting a new future to `this.incoming` in case of an error.

Reply:
Ah, thanks. I didn't know that.

> so there is no point in setting a new future

That was exactly my point =)